uaclient/api/u/pro/packages/updates_with_cves/v1.py

```diff
@@ -0,0 +1,353 @@
+            doc="Detailed list of all available updates",
+        ),
+        Field(
+            "cves",
+            data_list(CVEInfo),
+            doc="A list of CVEs that affect the system",
+        ),
+        Field(
+            "vulnerability_data_published_at",
+            DatetimeDataValue,
+            doc="The date the JSON vulnerability data was published at",
+        ),
+    ]
+
+    def __init__(
+        self,
+        summary: UpdateSummary,
+        updates: List[UpdateInfoWithCVES],
+        cves: List[CVEInfo],
+        vulnerability_data_published_at: datetime.datetime,
+    ):
+        self.summary = summary
+        self.updates = updates
+        self.cves = cves
+        self.vulnerability_data_published_at = vulnerability_data_published_at
+
+
+def updates_with_cves(
+    options: UpdatesInfoWithCVEsOptions,
+) -> PackageUpdatesWithCVEsResult:
+    return _updates_with_cves(options=options, cfg=UAConfig())
+
+
+def _get_pkg_current_version(cache, pkg: str):
+    return cache[pkg].current_ver.ver_str
+
+
+def _get_pkg_updates(cfg: UAConfig, updates_data: Optional[str]):
+    if not updates_data:
+        return _updates(cfg)
+
+    with open(updates_data, "r") as f:
+        package_updates = json.loads(f.read())
+        return PackageUpdatesResult.from_dict(
+            package_updates["data"]["attributes"]
+        )
+
+
+def _updates_with_cves(
+    options: UpdatesInfoWithCVEsOptions, cfg: UAConfig
+) -> PackageUpdatesWithCVEsResult:
+    """
+    This endpoint shows available updates for packages in a system including
+    the CVEs that are tied to each update.
+    """
+
+    package_updates = _get_pkg_updates(cfg, options.updates_data)
+    vulnerabilities_json_data = VulnerabilityData(
+        cfg=cfg, data_file=options.data_file, series=options.series
+    ).get()
+    package_updates_with_cves = []
+    cves_info = []
+    all_cves = set()
+
+    with PreserveAptCfg(get_apt_pkg_cache) as cache:
+        for pkg in package_updates.updates:
+            related_cves = []
+
+            pkg_name = pkg.package
+
+            source_pkg = _get_source_package_from_vulnerabilities_data(
+                vulnerabilities_json_data, pkg_name
+            )
+            update_version = pkg.version
+            current_pkg_version = _get_pkg_current_version(cache, pkg_name)
+
+            pkg_cves = (
+                vulnerabilities_json_data.get("packages")
+                .get(source_pkg, {})
+                .get("cves", {})
+            )
+
+            for cve_name, cve_info in pkg_cves.items():
+                cve_fixed_version = cve_info["source_fixed_version"]
+                if not cve_fixed_version:
+                    continue
+
+                if version_compare(current_pkg_version, cve_fixed_version) < 0:
+                    if version_compare(update_version, cve_fixed_version) >= 0:
+                        related_cves.append(cve_name)
+                        all_cves.add(cve_name)
```

Comment:
@panlinux @basak @dodys please take a look on the logic where we say if a CVE will be fixed by a package upgrade. I just want to be sure I am not missing anything here

Reply:
@lucasmoura 
`current_pkg_version` seems to be storing the binary pkg version right? Not the source pkg version
And then `cve_fixed_version` is storing the `source_fixed_version`.
If that is indeed what it is doing, then you will face issues, as binary pkgs might have a different version that its source. You might want to either compare source version against source version or compare binary version against binary version.

Reply:
@dodys fixed